### PR TITLE
Do not recurse for ExprKind::Paren lowering.

### PR DESCRIPTION
**Pull request by cjgillot**
In order to reduce stack consumption.

### CODE REVIEW

**Comment by Mark-Simulacrum** [comment]
I think this seems reasonable, in theory, but it might be better to ask if the paren expr should be not recursively added so eagerly if this is indeed an improvement? That has probably higher cost (allocations etc) than the decomposition done here... but maybe I'm missing something.

r? @petrochenkov

**Comment by cjgillot** [comment]
Sorry, I do not understand your comment @Mark-Simulacrum.

**Comment by Mark-Simulacrum** [comment]
I am not personally aware of the concrete uses of ExprKind::Paren, but it seems like this PR implies that it is common to see them directly nested many times. That suggests that it might be best to add a helper or so to perform this compaction when adding the nesting, rather than decomposing it during lowering.

**Comment by petrochenkov** [comment]
>this PR implies that it is common to see them directly nested many times

This is something I would like to confirm as well.
Do nested parentheses actually happen in practice? Maybe in macros?
@cjgillot , did you have some specific case in mind when implementing this? Could stacker help this case instead?

>That suggests that it might be best to add a helper or so to perform this compaction when adding the nesting

This may be pretty annoying given that e.g. macro calls in parentheses can expand to another pair of parentheses, also you need to maintain a stack of spans for each pair of parentheses.
In other words, doing this at lowering stage is at least much simpler.

**Comment by cjgillot** [comment]
@petrochenkov this seemed like a low-hanging fruit to reduce stack consumption.
I initially implemented this because the recursion annoyed me in https://github.com/rust-lang/rust/pull/79519.
I don't have a specific test case, except the one which prompted the initial use of `ensure_sufficient_stack`.

**Comment by petrochenkov** [comment]
As I remember, the general compiler team stance was to rely on stacker since there are numerous cases of recursion like this, and in most of them the "recursion -> loop" transformation makes code more complex, so closing.
